### PR TITLE
User can view shipping countries from pledge screen

### PR DESCRIPTION
**Pull request by Rcureton**
# What ❓
- This was a great collaboration between the three of us! @Scollaco @eoji 🎉
- You can now view shipping countries from the pledge screen for rewards with shipping rules.
- This is part of the `native checkout` feature. 

# How to QA? 🤔
- [ ] Go to a project with the `native checkout feature flag` enabled -> select a reward -> on the pledge screen 
- [ ]  If a reward has shipping rules then you will see the EditText which you can start typing the country you want to ship to.
- [ ] If a reward has no shipping rules then the entire row will be gone.

# Story 📖
[User can view shipping countries from pledge screen](https://trello.com/c/ClAhqsC2/1271-user-can-view-shipping-countries-from-pledge-screen-android)

# See 👀
| With Shipping Rules | Without Shipping Rules |
| --- | --- |
|  |  |

Apr-25-2019 16-48-06

### CODE REVIEW

**Comment by Scollaco** [comment]
[User Can View Shipping Countries from Pledge Screen [Android]](https://trello.com/c/ClAhqsC2/1271-user-can-view-shipping-countries-from-pledge-screen-android)

File: app/src/main/java/com/kickstarter/mock/factories/ShippingRuleFactory.kt [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+                    .cost(30.0)
+                    .location(LocationFactory.unitedStates())
+                    .build()
+        }
+
+        fun germanyShippingRule(): ShippingRule {
+            return ShippingRule.builder()
+                    .id(10)
+                    .cost(30.0)
+                    .location(LocationFactory.unitedStates())
```
Comment (Scollaco) [inline app/src/main/java/com/kickstarter/mock/factories/ShippingRuleFactory.kt]: Change the Location to germany

File: app/src/main/java/com/kickstarter/ui/fragments/PledgeFragment.kt [hunk 2]
```diff
@@ -304,6 +367,7 @@ class PledgeFragment : BaseFragment<PledgeFragmentViewModel.ViewModel>(), Reward
             rewardCardAdapter.resetSelectedPosition(position)
         }
         (cards_recycler.layoutManager as FreezeLinearLayoutManager).setFrozen(selected)
+        shipping_rules.dismissDropDown()
```
Comment (Scollaco) [inline app/src/main/java/com/kickstarter/ui/fragments/PledgeFragment.kt]: Move this to the appropriated place

File: app/src/main/java/com/kickstarter/ui/viewholders/KSArrayViewHolder.kt [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+
+    private val lifecycle = PublishSubject.create<ActivityEvent>()
+    abstract fun bindData(any: Any?)
+
+    override fun lifecycle(): Observable<ActivityEvent> {
+        return this.lifecycle
+    }
+
+    /**
+     * This method is intended to be called only from `KSAdapter` in order for it to inform the view holder
```
Comment (Scollaco) [inline app/src/main/java/com/kickstarter/ui/viewholders/KSArrayViewHolder.kt]: Copy pasta here

File: app/src/main/res/values/styles.xml [hunk 4]
```diff
@@ -698,5 +699,13 @@
     <item name="cardTint">@color/ksr_dark_grey_500</item>
     <item name="cardTextErrorColor">@color/ksr_red_400</item>
   </style>
+  
+  <!-- AutoComplete -->
+  <style name="AutocompleteStyle" parent="Widget.AppCompat.Light.AutoCompleteTextView">
+    <item name="android:colorControlActivated">@color/ksr_green_500</item>
+    <item name="android:background">@drawable/rect_white_rounded</item>
+    <item name="android:state_pressed">true</item>
```
[the remark this answers is not in the document]
Comment (Rcureton) [reply app/src/main/res/values/styles.xml]: I removed this I thought it handled the pressed state for the dropdown menu. Fixed this!

**Comment by Rcureton** [comment]
Updated screenshots with conversions and non conversion

File: app/src/main/java/com/kickstarter/viewmodels/PledgeFragmentViewModel.kt [hunk 5]
```diff
@@ -206,19 +269,83 @@ interface PledgeFragmentViewModel {
...
+                    .compose<Pair<Double, Project>>(combineLatestPair(project))
+                    .map<SpannableString> { this.ksCurrency.formatWithProjectCurrency(it.first, it.second, RoundingMode.UP, 2) }
+                    .compose(bindToLifecycle())
+
+            Observable.merge(initialTotalAmount, totalWithShippingRule)
+                    .compose(bindToLifecycle())
+                    .subscribe(this.totalAmount)
+
+            val projectAndReward = project
+                    .compose<Pair<Project, Reward>>(combineLatestPair(reward))
```
Comment (Rcureton) [inline app/src/main/java/com/kickstarter/viewmodels/PledgeFragmentViewModel.kt]: move to top and re-use it

File: app/src/main/java/com/kickstarter/viewmodels/PledgeFragmentViewModel.kt [hunk 6]
```diff
@@ -271,10 +410,22 @@ interface PledgeFragmentViewModel {
...
         override fun animateRewardCard(): Observable<PledgeData> = this.animateRewardCard
 
+        @NonNull
         override fun cards(): Observable<List<StoredCard>> = this.cards
 
+        @NonNull
         override fun continueButtonIsGone(): Observable<Boolean> = this.continueButtonIsGone
 
+        @NonNull
+        override fun conversionTextViewIsGone(): Observable<Boolean> {
```
Comment (Rcureton) [inline app/src/main/java/com/kickstarter/viewmodels/PledgeFragmentViewModel.kt]: change these to `=`

File: app/src/main/res/layout/fragment_pledge.xml [hunk 7]
```diff
@@ -172,19 +172,73 @@
...
           android:layout_marginBottom="@dimen/grid_4"
-          tools:text="$20" />
+          android:gravity="center_vertical"
+          android:orientation="horizontal"
+          android:weightSum="1.8">
+
+          <AutoCompleteTextView
+            android:id="@+id/shipping_rules"
+            style="@style/AutocompleteStyle"
+            android:layout_width="140dp"
```
Comment (Rcureton) [inline app/src/main/res/layout/fragment_pledge.xml]: fix dimen

File: app/src/main/res/layout/fragment_pledge.xml [hunk 8]
```diff
@@ -172,19 +172,73 @@
...
+
+            <View
+              android:id="@+id/shipping_amount_loading_view"
+              android:layout_width="@dimen/grid_8"
+              android:layout_height="@dimen/grid_2"
+              android:layout_gravity="center|end"
+              android:background="@drawable/pledge_amounts_loading_states" />
+          </FrameLayout>
+
+
```
Comment (Rcureton) [inline app/src/main/res/layout/fragment_pledge.xml]: nl

File: app/src/main/res/values/strings.xml [hunk 9]
```diff
@@ -45,6 +45,7 @@
   <string name="Not_implemented_yet">Not implemented yet, coming soon!</string>
   <string name="Open_downloads_folder">Open Downloads folder</string>
   <string name="playground_playground">Playground</string>
+  <string name="plus_sign">+</string>
```
Comment (Rcureton) [inline app/src/main/res/values/strings.xml]: remove

File: app/src/main/res/values/styles.xml [hunk 10]
```diff
@@ -698,5 +699,14 @@
     <item name="cardTint">@color/ksr_dark_grey_500</item>
     <item name="cardTextErrorColor">@color/ksr_red_400</item>
   </style>
+  
+  <!-- AutoComplete -->
+  <style name="AutocompleteStyle" parent="Widget.AppCompat.Light.AutoCompleteTextView">
+    <item name="android:colorControlActivated">@color/ksr_green_500</item>
+    <item name="android:background">@drawable/rect_white_rounded</item>
+    <item name="android:dropDownWidth">265dp</item>
```
Comment (Rcureton) [inline app/src/main/res/values/styles.xml]: add dimen

File: app/src/main/res/layout/fragment_pledge.xml [hunk 11]
```diff
@@ -172,19 +172,71 @@
...
+          <Space
+            android:layout_width="0dp"
+            android:layout_height="wrap_content"
+            android:layout_weight="1" />
+
+          <androidx.appcompat.widget.AppCompatImageView
+            android:layout_width="@dimen/plus_sign_width"
+            android:layout_height="@dimen/plus_sign_height"
+            android:backgroundTint="@color/ksr_grey_500"
+            android:src="@drawable/ic_add"
```
Comment (eoji) [inline app/src/main/res/layout/fragment_pledge.xml]: It's too small and it's the wrong color.

Comment (Rcureton) [reply app/src/main/res/layout/fragment_pledge.xml]: I fixed the color! Also the abstract has properties of `20dp for height` and `9dp for width`

https://app.goabstract.com/projects/ff85f7f0-23b6-11e8-9c12-3d345000cd85/branches/master/commits/4cd104da8a116335c9a2861a888a0bdba50a261d/files/DD4807BD-4D34-43F3-853E-4765774306D1/layers/426005EF-46B4-47F2-B95E-45C092D754AB?mode=build&nonce=2887376e-9c5f-4696-b33b-f0729f16e0bf&selected=3355706653-61DBAB2D-6F81-458E-B42B-3B6C7EA9854C

Comment (eoji) [reply app/src/main/res/layout/fragment_pledge.xml]: A drawable has intrinsic padding so when you said a fixed size and not a scaletype, it doesn't take that into account. It's too small.

Comment (eoji) [reply app/src/main/res/layout/fragment_pledge.xml]: The positioning is also off.

**Comment by eoji** [comment]
I took another look and I can still reproduce both bugs.

1. The keyboard bug is still happening and the 500 millisecond delay is very noticeable. The delay also happens even if the keyboard is not up so it just makes the app seem slow.
device-2019-05-31-142450 2019-05-31 14_31_09

2. The plus sign is still off. You can notice it especially when both are visible on the screen.


I would very much like to work on the checkout mutation this sprint. I think we should timebox these bugs to Tuesday EOD so this PR doesn't stay open any longer.

**Pull request merged**